Diagnostic renaming 3

nnethercote opened the pull request:
A sequel to https://github.com/rust-lang/rust/pull/121780.

r? @davidtwco

nnethercote committed:
Rename `IntoDiagnosticArg` as `IntoDiagArg`.

Also rename `into_diagnostic_arg` as `into_diag_arg`, and
`NotIntoDiagnosticArg` as `NotInotDiagArg`.

nnethercote committed:
Rename `IntoDiagnostic` as `Diagnostic`.

To match `derive(Diagnostic)`.

Also rename `into_diagnostic` as `into_diag`.

nnethercote committed:
Rename `AddToDiagnostic` as `Subdiagnostic`.

To match `derive(Subdiagnostic)`.

Also rename `add_to_diagnostic{,_with}` as `add_to_diag{,_with}`.

nnethercote committed:
Remove `Default` derive from `MacroExpectedFound`.

It's the only diagnostic in the entire project that impls `Default`, and
the code is clearer without it.

nnethercote committed:
Remove unused impls for some diagnostics.

nnethercote committed:
Remove unused impls from some subdiagnostics.

davidtwco commented on compiler/rustc_ast_lowering/src/errors.rs:
```diff
@@ -1,10 +1,10 @@
 use rustc_span::{symbol::Ident, Span, Symbol};
 
-#[derive(Diagnostic)]
+#[derive(IntoDiag)]
```
Does it make sense to go this way or the opposite and rename the traits? `#[derive(Diag)]` and `trait Diag` might be better than `#[derive(IntoDiag)]` and `trait IntoDiag`?

nnethercote replied:
Hmm, good suggestion. We'd have `derive(Diag)`, `derive(Subdiag)`, `derive(LintDiag)`. I'll do that.

nnethercote replied:
Turns out that doesn't work because `Diag` and `Subdiag` are already names of types, so they can't also be names of traits. Instead I'll revert to the original `derive(Diagnostic)`, etc., and change the trait names to match.

davidtwco reviewed (approved):
LGTM, one suggestion, can be ignored or a follow-up or a change here.

nnethercote committed:
Rename `DecorateLint` as `LintDiagnostic`.

To match `derive(LintDiagnostic)`.

nnethercote committed:
Rename diagnostic derive things.

For increased consistency.

- session_diagnostic_derive    -> diagnostic_derive
- session_subdiagnostic_derive -> subdiagnostic_derive
- SubdiagnosticDeriveBuilder   -> SubdiagnosticDerive

nnethercote commented:
New code is up, with some sizeable changes so I won't just r+ it myself.

Now `derive(Diagnostic)`/`derive(Subdiagnostic)`/`derive(LintDiagnostic)` are unchanged, but the corresponding trait names are changed to match the derive attributes:
- `IntoDiagnostic` -> `Diagnostic`
- `AddToDiagnostic` -> `Subdiagnostic`
- `DecorateLint` -> `LintDiagnostic`

This also removes the need to rename various types, e.g. `DiagnosticDeriveKind` shouldn't be renamed `DiagDeriveKind` given that the trait/derive name is `Diagnostic`.

Overall this makes the change quite a bit smaller, which is nice.

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit b91031ae1939641e2c8efd5dd35510d64e875910 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco b91031ae1939641e2c8efd5dd35510d64e875910 -->
<!-- homu: {"type":"Approved","sha":"b91031ae1939641e2c8efd5dd35510d64e875910","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #122182) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

@bors r+ p=1

Because this is conflict-prone.

bors commented:
:pushpin: Commit 43c06335cf2d1cc6b12fcfa5b8795d9becc52e8c has been approved by `nnethercote`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nnethercote 43c06335cf2d1cc6b12fcfa5b8795d9becc52e8c -->
<!-- homu: {"type":"Approved","sha":"43c06335cf2d1cc6b12fcfa5b8795d9becc52e8c","approver":"nnethercote","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 43c06335cf2d1cc6b12fcfa5b8795d9becc52e8c with merge cd81f5b27ee00b49d413db50b5e6af871cebcf23...
<!-- homu: {"type":"BuildStarted","head_sha":"43c06335cf2d1cc6b12fcfa5b8795d9becc52e8c","merge_sha":"cd81f5b27ee00b49d413db50b5e6af871cebcf23"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8225932022/job/22494121502)
Approved by: nnethercote
Pushing cd81f5b27ee00b49d413db50b5e6af871cebcf23 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nnethercote","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8225932022/job/22494121502"},"merge_sha":"cd81f5b27ee00b49d413db50b5e6af871cebcf23"} -->

rust-timer commented:
Finished benchmarking commit (cd81f5b27ee00b49d413db50b5e6af871cebcf23): [comparison URL](https://perf.rust-lang.org/compare.html?start=76ee6fc926032c16a5706086922fc1125c061b70&end=cd81f5b27ee00b49d413db50b5e6af871cebcf23&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=76ee6fc926032c16a5706086922fc1125c061b70&end=cd81f5b27ee00b49d413db50b5e6af871cebcf23&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=76ee6fc926032c16a5706086922fc1125c061b70&end=cd81f5b27ee00b49d413db50b5e6af871cebcf23&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.8%, -2.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-3.2%, -1.5%] | 10    |
| All ❌✅ (primary)                 | -2.6% | [-2.8%, -2.2%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=76ee6fc926032c16a5706086922fc1125c061b70&end=cd81f5b27ee00b49d413db50b5e6af871cebcf23&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=76ee6fc926032c16a5706086922fc1125c061b70&end=cd81f5b27ee00b49d413db50b5e6af871cebcf23&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 646.276s -> 648.469s (0.34%)
**Artifact size**: 309.98 MiB -> 309.94 MiB (-0.01%)
